sdk/python/lib/pulumi/resource.py

```diff
@@ -49,6 +53,160 @@ def __init__(self,
         self.delete = delete
 
 
+def inherited_child_alias(
+        child_name: str,
+        parent_name: str,
+        parent_alias: 'Input[str]',
+        child_type: str) -> 'Output[str]':
+    """
+    inherited_child_alias computes the alias that should be applied to a child based on an alias
+    applied to it's parent. This may involve changing the name of the resource in cases where the
+    resource has a named derived from the name of the parent, and the parent name changed.
+    """
+
+#   If the child name has the parent name as a prefix, then we make the assumption that it was
+#   constructed from the convention of using `{name}-details` as the name of the child resource.  To
+#   ensure this is aliased correctly, we must then also replace the parent aliases name in the prefix of
+#   the child resource name.
+#
+#   For example:
+#   * name: "newapp-function"
+#   * opts.parent.__name: "newapp"
+#   * parentAlias: "urn:pulumi:stackname::projectname::awsx:ec2:Vpc::app"
+#   * parentAliasName: "app"
+#   * aliasName: "app-function"
+#   * childAlias: "urn:pulumi:stackname::projectname::aws:s3/bucket:Bucket::app-function"
+    from .output import Output as Op
+    alias_name = Op.from_input(child_name)
+    if child_name.startswith(parent_name):
+        alias_name = Op.from_input(parent_alias).apply(
+            lambda u: u[u.rfind("::") + 2:] + child_name[len(parent_name):])
+
+    return create_urn(alias_name, child_type, parent_alias)
+
+ABSENT_VALUE = object()
+"""
+ABSENT_VALUE is how we mark if a value provided to an alias was actually given by the user as 'None',
+versus the use not providing any value at all.
+"""
+
+ROOT_STACK_RESOURCE = None
+"""
+Constant to represent the 'root stack' resource for a Pulumi application.  The purpose of this is
+solely to make it easy to write an [Alias] like so:
+
+`aliases=[Alias(parent=pulumi.ROOT_STACK_RESOURCE)]`.
+
+This indicates that the prior name for a resource was created based on it being parented directly by
+the stack itself and no other resources.  Note: this is equivalent to:
+
+`aliases=[Alias(parent=None)]`
+
+However, the former form is preferable as it is more self-descriptive, while the latter may look a
+bit confusing and may incorrectly look like something that could be removed without changing
+semantics.
+"""
+
+class Alias:
+    """
+    Alias is a partial description of prior named used for a resource. It can be processed in the
+    context of a resource creation to determine what the full aliased URN would be.
+
+    Note there is a semantic difference between attributes being given the `None` value and
```

Comment:
That's a good point.  @astronouth7303 this is how we're doing things in JS/TS.  it's subtle, but it allows callers to say "just use the default" by just doing something like `Alias(name="whatever")` versus say "actually use None" by doing somethign like `Alias(name="whatever", parent=None)`.

Note that aliases are an extremely advanced library maintenance scenario.  They're not expected for people to normally use.  So i'm personally ok with it being somewhat complex.

Reply:
While this can and does happen in Python, it's usually done with a default value in the signature just because that's the easiest way to go.

In this situation, there are three values: "Look up the current parent", "Don't have a parent at all", and "Use this specific object as the parent".

This not a common opinion, but one I rather like: Use the literal `...` to mean the first option. (The use of `...` generally to mean `Ellipsis` was new in 3.0). So my suggestion is to replace `ABSENT_VALUE` in all of this with `...` to mean "Please compute a reasonable value".

So the second example could also be written as `Alias(name="whatever", parent=...)`.

Reply:
To be clear, the common "absent value" sentinel is `None`. But when `None` has a semantic meaning and the default is to calculate/look up some value, I like to use `...` because it kinda feels like "ehh... fill in the blank plzkthx".

Reply:
Ok.  I think i get it.  Let me try this out :)

Reply:
Nice!  This worked.  Thanks @astronouth7303.

I'm going to merge in in this form.  If you think there are other appropriate changes we should make to the general API surface area here, let me know!